internal/config/image/cuda_image.go

```diff
@@ -219,11 +234,14 @@ func (i CUDA) OnlyFullyQualifiedCDIDevices() bool {
 // VisibleDevicesFromEnvVar returns the set of visible devices requested through
 // the NVIDIA_VISIBLE_DEVICES environment variable.
 func (i CUDA) VisibleDevicesFromEnvVar() []string {
+	if i.IsSwarmResource() {
+		return i.DevicesFromEnvvars(i.visibleDevicesEnvVars...).List()
+	}
```

Comment:
Considering how this is implemented in `container_config`, I believe this should be:
```
	for _, envvar := range i.visibleDevicesEnvVars {
		if _, ok := i.env[envvar]; ok {
			return i.DevicesFromEnvvars(envvar).List()
		}
	}
```
(I think there was a bug in the original implementation, but I will have to confirm).